keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,431 @@
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+The following metrics will be reported for pods that are in the `Running` phase and follow the same rules as above, but are not reported when the pod is `Pending` and can be used to calculate live resources. They all share the form `kube_running_pod_resource_(requests|limits)[_<resource>_<unit>]`.
+
+* `kube_running_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time. It will always report the `node` label since running pods must be assigned to a node.
+* `kube_running_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time. It will always report the `node` label since running pods must be assigned to a node.
+
+This will add `O(pods * resource types)` metrics series, and in general most clusters have 4-8 resource types. The series per pod is `non_zero_resource_count * 2 * 2 (requests and limits for normal and _lifecycle)`.  This explicitly does not expose containers as a dimension, as the resource model applies rules to containers that cannot be trivially expressed via a formula and does not help answer the core questions.
+
+These metrics should roughly correspond to:
+
+* The resources the scheduler will consider the pod to be consuming at the current time
+* The resources the kubelet will use to decide whether to admit the pod (nodes will reject requested resources that exceed their actual allocatable capacity)
+* The effective reserved capacity as resource quota would observe for the namespace
+* The sum of all resources consumed on the cluster
+* The size of cgroup constraints applied to pods on a node
+
+The lifecycle metrics allow key insights that are impossible today:
+
+* Observing when init containers or containers have substantially different resource reservation, causing workloads to under provision nodes
+* Visualize batch compute consumption on the cluster where completed pods can be quickly excluded
+
+The metrics would be served by the scheduler that is holding the active scheduler lease. During a failover, no metrics would be reported. This would prevent duplicate metrics from being scraped and aligns with our decision in the scheduler to not fill caches prior to election for resource usage minimization.
+
+Custom schedulers based on the scheduler code should be able to easily reuse and report these metrics, and the configuration of the cluster instrumentation to retrieve those metrics would be left as an exercise for the integrator.
```

Comment:
In terms of scheduler implementation, as current KEP suggested, scheduler will need to expose the pod's metric right after the binding phase - that's the time frame scheduler "assume"s a pod's scheduling, this is pluggable as this work can be accomplished in the PostFilter extension point.

However, once a Pod gets deleted, or a Job's Pod turns to Completed state, currently scheduler simply remove that pod's res usage from its internal cache. My question is: is scheduler also responsible for unhook the usage of that deleted/completed pod? (if so, this logic is not pluggable)

Reply:
Will clarify (since I think this confused others) in the KEP how the model works.  Effectively we iterate over the pod cache and report metrics from the current snapshot, so no integration into the internal scheduler state is necessary.